### PR TITLE
Content collections api reference

### PR DESCRIPTION
**Pull request by bholmesdev**
#### What kind of changes does this PR include?

- New or updated content

#### Description

Add content collections API reference. This documents all properties in `astro:content`: the `CollectionEntry` type, `defineCollection()`, `getCollection()`, and `getEntry()`.

Also considered an entry in "configuration" for the `src/content/config.*`. Since this is experimental and Configuration is scoped to `astro.config`, I left this for a future PR!

### CODE REVIEW

File: src/pages/en/reference/api-reference.md [hunk 1]
```diff
@@ -701,6 +701,149 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
...
+const allBlogPosts = await getCollection('blog');
+
+// Only return posts with `draft: true` in the frontmatter
+const draftBlogPosts = await getCollection('blog', ({ data }) => {
+  return data.draft === true;
+});
+---
+```
+
+[See the `getCollection()` guide section](/en/guides/content-collections#getcollection) for complete usage examples.
```
Comment (sarah11918) [inline src/pages/en/reference/api-reference.md:792]:
```suggestion
[See the `getCollection()` guide section](/en/guides/content-collections/#getcollection) for complete usage examples.
```

File: src/pages/en/reference/api-reference.md [hunk 2]
```diff
@@ -701,6 +701,149 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
 | `page.url.prev`    | `string \| undefined` | Get the URL of the previous page (will be `undefined` if on page 1).                                                              |
 | `page.url.next`    | `string \| undefined` | Get the URL of the next page (will be `undefined` if no more pages).                                                              |
 
+## Content Collections (Experimental)
+
+<p>
+  <Since v="1.7.0" />
+</p>
+
+Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections).
```
Comment (sarah11918) [inline src/pages/en/reference/api-reference.md:710]:
```suggestion
Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).
```

File: src/pages/en/reference/api-reference.md [hunk 3]
```diff
@@ -701,6 +701,149 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
...
+
+#### `slug()`
+
+**Type:** `(entry: { defaultSlug: string } & Omit<CollectionEntry, 'slug'>) => string | undefined`
+
+`slug()` is an optional function to override the default entry slug generated by Astro. This function receives the `defaultSlug`, entry `id`, unparsed document `body`, and `data` parsed by your configured schema, if any.
+
+#### Usage
+
+Say you have a collection named `src/content/blog/`, where each Markdown entry has a title and optional `permalink`. You can use `defineCollection` to enforce this frontmatter with the `schema` property. Then, you can map the `permalink` to the entry's `slug` if it exists:
```
Comment (sarah11918) [inline src/pages/en/reference/api-reference.md:829]:
```suggestion
```
Like all the other times a code sample is used on this page, I would tend not to give this any preamble or explain what it's doing. The guide page shows how to define a collection, enforce frontmatter properties and customize the slug. The code samples here can just be "what it looks like" without much commentary.

I do like the code comment for the export! I think that's sufficient for what needs to go here!

**Review by sarah11918 — COMMENTED** [review]
Yay @bholmesdev! What a way to end the year!  🥳 

I made a few small suggestions/call outs from my read through, mostly for consistency etc. I think you nailed the "what needs to go on this page" vs what can be left to the guide page in docs, so this is just a tiny bit of clean up!

I'm going to let Chris take a pass, too, and then might read fresh again tomorrow, especially if there are any more changes made.

File: src/pages/en/reference/api-reference.md [hunk 4]
```diff
@@ -701,6 +701,149 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
...
 
+## Content Collections (Experimental)
+
+<p>
+  <Since v="1.7.0" />
+</p>
+
+Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).
+
+### Collection Entry Type
```
Comment (sarah11918) [inline src/pages/en/reference/api-reference.md:712]: question: Does it maybe make sense to \show these closer to the order in which someone might encounter/use them?

It almost feels like burying the lede a bit to save "defining a collection" to the very end? I might be tempted to go `defineCollection()` first. Then I'd probably think of `getCollection()` and `getEntry()` first, before the the Collection Entry Type? That way, we also don't have the reference of "things below yet to come" in this first line. That is often a sign that maybe things should go in a different order.

I know for this reference page it's not as important as our Guide page in docs. But, I still kind of like organizing these in the order someone might use, and therefore understand, them. Someone *might* be reading these in order, and that seems like a logical flow to me.

File: src/pages/en/reference/api-reference.md [hunk 5]
```diff
@@ -701,6 +701,149 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
...
+
+**Type:** `(entry: { defaultSlug: string } & Omit<CollectionEntry, 'slug'>) => string | undefined`
+
+`slug()` is an optional function to override the default entry slug generated by Astro. This function receives the `defaultSlug`, entry `id`, unparsed document `body`, and `data` parsed by your configured schema, if any.
+
+#### Usage
+
+Say you have a collection named `src/content/blog/`, where each Markdown entry has a title and optional `permalink`. You can use `defineCollection` to enforce this frontmatter with the `schema` property. Then, you can map the `permalink` to the entry's `slug` if it exists:
+
+```ts
```
Comment (sarah11918) [inline src/pages/en/reference/api-reference.md:831]:
```suggestion
```ts
// src/content/config.ts
import { z, defineCollection } from 'astro:content';
```
We don't show importing `defineCollection()` here, though we do show exporting it. Is this helpful?

**Review by sarah11918 — COMMENTED** [review]
A couple more thoughts!

File: src/pages/en/reference/api-reference.md [hunk 6]
```diff
@@ -701,6 +701,149 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
 | `page.url.prev`    | `string \| undefined` | Get the URL of the previous page (will be `undefined` if on page 1).                                                              |
 | `page.url.next`    | `string \| undefined` | Get the URL of the next page (will be `undefined` if no more pages).                                                              |
 
+## Content Collections (Experimental)
+
+<p>
+  <Since v="1.7.0" />
+</p>
+
+Content collections offer APIs to configure and query your Markdown documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).
```
[the remark this answers is not in the document]
Comment (bholmesdev) [reply src/pages/en/reference/api-reference.md]: No prob! Adding "or MDX"

File: src/pages/en/reference/api-reference.md (hunk 4 above)
Comment (bholmesdev) [reply src/pages/en/reference/api-reference.md]: Agreed! Reordering to:
1. `defineCollection`
2. `getCollection` and `getEntry`
3. Collection type

**Comment by bholmesdev** [comment]
@sarah11918 Edits made! I also moved the usage example to the top of `defineCollection()` like so with the preamble removed. Let me know if this change makes sense. [Preview link](https://deploy-preview-2217--astro-docs-2.netlify.app/en/reference/api-reference/#definecollection)

image

**Comment by sarah11918** [comment]
I love it, Ben!  🥳

File: src/pages/en/reference/api-reference.md [hunk 7]
```diff
@@ -701,6 +701,147 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
 | `page.url.prev`    | `string \| undefined` | Get the URL of the previous page (will be `undefined` if on page 1).                                                              |
 | `page.url.next`    | `string \| undefined` | Get the URL of the next page (will be `undefined` if no more pages).                                                              |
 
+## Content Collections (Experimental)
+
+<p>
+  <Since v="1.7.0" />
+</p>
+
+Content collections offer APIs to configure and query your Markdown or MDX documents in `src/content/`. For feature rundowns and usage examples, [see our content collections guide](/en/guides/content-collections/).
```
Comment (sarah11918) [inline src/pages/en/reference/api-reference.md:710]:
```suggestion
Content collections offer APIs to configure and query your Markdown or MDX documents in `src/content/`. For features and usage examples, [see our content collections guide](/en/guides/content-collections/).
```
nit: Now that the example pulled my attention back up here, I think we can go a little less colloquial on the API reference page.

But, I think.... after that we can update branch?? 🎉

File: src/pages/en/reference/api-reference.md [hunk 8]
```diff
@@ -701,6 +701,147 @@ Pagination will pass a `page` prop to every rendered page that represents a sing
...
+
+#### `render()`
+
+**Type:** `() => Promise<RenderedEntry>`
+
+A function to compile a given Markdown or MDX document for rendering. This returns the following properties:
+
+- `<Content />` - A component used to render the document's contents in an Astro file.
+- `headings` - A generated list of headings, [mirroring Astro's `getHeadings()` utility](/en/guides/markdown-content/#exported-properties) on Markdown and MDX imports.
+- `injectedFrontatter` - An object of frontmatter [injected via remark or rehype plugins](/en/guides/markdown-content/#example-injecting-frontmatter). Set to type `any`.
```
Comment (chenxsan) [inline src/pages/en/reference/api-reference.md:832]:
```suggestion
- `injectedFrontmatter ` - An object of frontmatter [injected via remark or rehype plugins](/en/guides/markdown-content/#example-injecting-frontmatter). Set to type `any`.
```

**Review by sarah11918 — APPROVED** [review]

**Pull request merged**